File: backend/routes/banners.ts
import { Router, Request, Response } from 'express';
import db from '../config/database';
import { rewriteRowImageFields } from '../config/cdn';

const router = Router();

// GET /api/banners — public, returns active banners ordered for the carousel.
// image_url is rewritten from raw S3 to the public CDN.
router.get('/', async (_req: Request, res: Response) => {
  try {
    const result = await db.query(
      `SELECT id, image_url, link_url, alt_text
         FROM banners
        WHERE is_active = TRUE
        ORDER BY sort_order ASC, created_at DESC`,
    );
    const banners = result.rows.map(r => rewriteRowImageFields(r, { url: ['image_url'] }));
    return res.json({ success: true, banners });
  } catch (err) {
    return res.status(500).json({ success: false, error: (err as Error).message });
  }
});

export default router;

File: backend-golang/internal/handlers/banners.go
package handlers

import (
	"context"
	"net/http"

	"inanhxink/backend-golang/internal/config"
)

// GET /api/banners — public, active banners ordered by sort_order.
// image_url is rewritten to CDN.
func ListBanners(w http.ResponseWriter, r *http.Request) {
	rows, err := config.DB.Query(context.Background(), `
		SELECT id, image_url, link_url, alt_text
		FROM banners
		WHERE is_active = TRUE
		ORDER BY sort_order ASC, created_at DESC`)
	if err != nil {
		InternalError(w, err)
		return
	}
	banners, err := CollectRows(rows)
	if err != nil {
		InternalError(w, err)
		return
	}
	for _, b := range banners {
		config.CdnURLField(b, "image_url")
	}
	OK(w, map[string]any{"success": true, "banners": banners})
}
